File: go/kernels.go
// Package batchinvariant is a Go port of the batch-invariant reduction kernels.
//
// A matmul output is a dot product: a sum over the K dimension. In exact
// arithmetic the order of that sum does not matter. In floating point it does —
// (a + b) + c can differ from a + (b + c) in the last bit, and far more when
// large terms nearly cancel. Production kernels split the K reduction into
// chunks and combine partial sums, and how many chunks is often chosen from the
// batch shape. So a row's result can change depending on who else is in the
// batch — a real source of non-reproducible LLM inference.
//
// SumSplit reduces by grouping; MatmulVariant chooses the grouping from the
// batch size (so a row moves with its neighbours), while MatmulInvariant fixes
// the reduction order regardless of batch (so a row is always the same bits).
package batchinvariant

// Vec is a dense vector of float64 values.
type Vec = []float64

// Mat is a dense matrix stored as a slice of rows.
type Mat = [][]float64

// FixedSplits is a single, batch-independent reduction shape. The value does
// not matter for correctness — only that it never depends on the batch.
const FixedSplits = 4

// SumFlat returns the left-to-right sum of vals — one fixed order.
func SumFlat(vals Vec) float64 {
	acc := 0.0
	for _, v := range vals {
		acc += v
	}
	return acc
}

// SumSplit sums vals by partitioning them into nsplits contiguous groups,
// reducing each group left-to-right, then summing the partials. It is
// mathematically equal to SumFlat; in floating point, a different grouping is a
// different result. This models a split-K / tiled reduction.
func SumSplit(vals Vec, nsplits int) float64 {
	n := len(vals)
	if nsplits <= 1 || n == 0 {
		return SumFlat(vals)
	}
	size := (n + nsplits - 1) / nsplits // ceil(n / nsplits)
	partials := make(Vec, 0, nsplits)
	for g := 0; g < nsplits; g++ {
		start := g * size
		if start >= n {
			continue
		}
		end := start + size
		if end > n {
			end = n
		}
		partials = append(partials, SumFlat(vals[start:end]))
	}
	return SumFlat(partials)
}

// SplitsForBatch picks the K-split count from the batch size (fewer rows → split
// K further to keep every core busy). Because the result depends on this number,
// it makes the kernel batch-variant. This models shape-dependent tiling that
// real kernels do.
func SplitsForBatch(batchSize int) int {
	d := batchSize
	if d < 1 {
		d = 1
	}
	s := 8 / d
	if s < 1 {
		s = 1
	}
	return s
}

func matmul(xBatch Mat, w Mat, nsplits int) Mat {
	out := make(Mat, 0, len(xBatch))
	for _, x := range xBatch {
		row := make(Vec, len(w))
		for j := range w {
			products := make(Vec, len(x))
			for i := range x {
				products[i] = x[i] * w[j][i]
			}
			row[j] = SumSplit(products, nsplits)
		}
		out = append(out, row)
	}
	return out
}

// MatmulVariant computes logits = x @ wᵀ, reduced with a batch-size-dependent
// grouping. xBatch is [batch][d_in], w is [d_out][d_in]. The K reduction uses
// SplitsForBatch(len(xBatch)) groups — so a row's output depends on how many
// rows share its batch. This is the bug.
func MatmulVariant(xBatch Mat, w Mat) Mat {
	return matmul(xBatch, w, SplitsForBatch(len(xBatch)))
}

// MatmulInvariant computes logits = x @ wᵀ with a reduction order fixed
// independently of the batch. Every dot product is reduced with the same
// grouping whatever the batch size, so a row's output is identical whether it
// runs alone or alongside a thousand others. That is the guarantee.
func MatmulInvariant(xBatch Mat, w Mat) Mat {
	return matmul(xBatch, w, FixedSplits)
}

File: ts/src/kernels.ts
/**
 * Reduction kernels — where batch-dependence sneaks in.
 *
 * A matmul output is a dot product: a sum over the K dimension. In exact
 * arithmetic the order of that sum does not matter. In floating point it does —
 * `(a + b) + c` can differ from `a + (b + c)` in the last bit, and far more when
 * large terms nearly cancel. Production kernels split the K reduction into
 * chunks and combine partial sums, and how many chunks is often chosen from the
 * batch shape, so a row's result can change depending on who else is in the
 * batch.
 */

/** A dense vector of float64 values. */
export type Vec = number[];

/** A dense matrix stored as an array of rows. */
export type Mat = number[][];

/**
 * A single, batch-independent reduction shape. The value does not matter for
 * correctness — only that it never depends on the batch.
 */
export const FIXED_SPLITS = 4;

/** Left-to-right sum of `vals` — one fixed order. */
export function sumFlat(vals: Vec): number {
  let acc = 0.0;
  for (const v of vals) {
    acc += v;
  }
  return acc;
}

/**
 * Sum `vals` by partitioning them into `nsplits` contiguous groups, reducing
 * each group left-to-right, then summing the partials. Mathematically equal to
 * {@link sumFlat}; in floating point, a different grouping is a different
 * result. This models a split-K / tiled reduction.
 */
export function sumSplit(vals: Vec, nsplits: number): number {
  const n = vals.length;
  if (nsplits <= 1 || n === 0) {
    return sumFlat(vals);
  }
  const size = Math.ceil(n / nsplits);
  const partials: Vec = [];
  for (let g = 0; g < nsplits; g++) {
    const start = g * size;
    if (start >= n) {
      continue;
    }
    const end = Math.min(start + size, n);
    partials.push(sumFlat(vals.slice(start, end)));
  }
  return sumFlat(partials);
}

/**
 * Pick the K-split count from the batch size (fewer rows → split K further to
 * keep every core busy). Because the result depends on this number, it makes the
 * kernel batch-variant.
 */
export function splitsForBatch(batchSize: number): number {
  const d = Math.max(1, batchSize);
  return Math.max(1, Math.floor(8 / d));
}

function matmul(xBatch: Mat, w: Mat, nsplits: number): Mat {
  return xBatch.map((x) =>
    w.map((wj) => {
      const products = x.map((xi, i) => xi * wj[i]);
      return sumSplit(products, nsplits);
    }),
  );
}

/**
 * logits = x @ wᵀ, reduced with a batch-size-dependent grouping. `xBatch` is
 * [batch][d_in], `w` is [d_out][d_in]. The K reduction uses
 * `splitsForBatch(xBatch.length)` groups — so a row's output depends on how many
 * rows share its batch. This is the bug.
 */
export function matmulVariant(xBatch: Mat, w: Mat): Mat {
  return matmul(xBatch, w, splitsForBatch(xBatch.length));
}

/**
 * logits = x @ wᵀ with a reduction order fixed independently of the batch. Every
 * dot product is reduced with the same grouping whatever the batch size, so a
 * row's output is identical whether it runs alone or alongside a thousand
 * others. That is the guarantee.
 */
export function matmulInvariant(xBatch: Mat, w: Mat): Mat {
  return matmul(xBatch, w, FIXED_SPLITS);
}
